liberica/src/lib/websockets.ts

```diff
@@ -1,49 +1,89 @@
-import { ClientMessage, GameState } from "lib/bindings";
+import { ClientMessage, ServerMessage } from "lib/bindings";
+
+export type Keys<T> = T extends T ? keyof T : never;
+export type Concrete<T, K extends Keys<T>> = T extends { [P in K]: infer V }
+  ? V
+  : never;
 
-export type WSHandler = (msg: GameState) => void;
 export type WSDiconnectHandler = () => void;
 export type WSErrorHandler = (error?: string | Error | object) => void;
+export type WSHandler<T, K extends Keys<T>> = (msg: Concrete<T, K>) => void;
+export type WSHandlerMap<M extends object> = {
+  [K in Keys<M>]?: WSHandler<M, K>;
+};
+
+export type WSEvent =
+  | { Disconnect: void }
+  | { Connect: void }
+  | { Error: unknown }; // Todo: Better error types
 
 export class WebsocketApi {
-  lastMessage?: Date;
-  connection: WebSocket;
-  handlers: WSHandler[];
-  errorHandler: WSErrorHandler;
-  disconnectHandler: WSDiconnectHandler;
+  public lastMessage?: Date;
 
-  constructor(endpoint: string, connected: (api: WebsocketApi) => void) {
-    this.connection = new WebSocket(endpoint);
+  private connection!: WebSocket;
+  private endpoint!: string;
 
-    this.handlers = [];
-    this.errorHandler = (x) => console.error("Websocket Error occured", x);
-    this.connection.onerror = this.errorHandler;
-    this.disconnectHandler = () => console.log("Closed WS connection");
-    this.connection.onopen = () => {
-      connected(this);
-      console.log("Established WS connection");
-    };
-    this.connection.onclose = () => {
-      this.disconnectHandler();
-    };
+  private handlers: WSHandlerMap<ServerMessage> = {};
+  private metaHandlers: WSHandlerMap<WSEvent> = {};
+
+  constructor(endpoint: string) {
+    this.connect(endpoint);
+  }
+
+  public reconnect(force: boolean = false) {
+    // Don't try to reconnect if there is a connection already
+    if (this.connection?.readyState === this.connection.OPEN && force) return;
+
+    this?.disconnect();
```

Comment:
does this do anything?

Reply:
It is possible, that the function is called from outside, where you want to force a disconnect.
And otherwise we want to prevent a reconnect if the current connection is already open.

And if we want to force a reconnect, then we also want to cleanly disconnect the already open connection.

But the `?`-Operator is actually useless, since this can never be falsy.